### PR TITLE
Added New Tests in Signum

### PR DESCRIPTION
**Pull request by imSanko**
### Describe your change:
Added new tests in Signum

* [ ] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [x] Add or change doctests? -- Note: Please avoid changing both code and tests in a single pull request.
* [ ] Documentation change?

### Checklist:
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms include at least one URL that points to Wikipedia or another similar explanation.
* [x] If this pull request resolves one or more open issues then the description above includes the issue number(s) with a [closing keyword](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

### CODE REVIEW

**Comment by cclauss** [comment]
Please do NOT remove tests.  The original developer might have had very good reasons for creating those particular tests.

Also, please run `python3 -m doctest -v maths/signum.py` on your own computer so you know if your proposed changes are going to pass the `build` test
```
=================================== FAILURES ===================================
________________________ [doctest] maths.signum.signum _________________________
007 
008      Applies signum function on the number
009 
010     Custom test cases:
011     >>> signum(-20)
Expected:
    -2
Got:
    -1
```
Finally, please try to add tests that show how the algorithm fails with difficult or garbage input.
```
     >>> signum(-21)
     >>> signum(1_000_000_000)
     >>> signum(12.345)
     >>> signum("a")
     >>> signum([])
```

**Review by cclauss — CHANGES_REQUESTED** [review]
Please bring back the old tests and add more.

File: maths/signum.py [hunk 1]
```diff
@@ -7,9 +7,10 @@ def signum(num: float) -> int:
     """
     Applies signum function on the number
 
-    >>> signum(-10)
```
Comment (cclauss) [inline maths/signum.py:10]: Please put this line back in place.

File: maths/signum.py [hunk 2]
```diff
@@ -7,9 +7,10 @@ def signum(num: float) -> int:
...
+    Custom test cases:
+    >>> signum(-20)
     -1
-    >>> signum(10)
```
Comment (cclauss) [inline maths/signum.py:12]: Please put this line back in place.

File: maths/signum.py [hunk 3]
```diff
@@ -26,9 +27,15 @@ def test_signum() -> None:
     assert signum(5) == 1
     assert signum(-5) == -1
     assert signum(0) == 0
+    assert signum(10.5) == 1
+    assert signum(-10.5) == -1
+    assert signum(1e-6) == 1
+    assert signum(-1e-6) == -1
+    assert signum(123456789) == 1
+    assert signum(-123456789) == -1
```
Comment (cclauss) [inline maths/signum.py:35]: Please add:
```
     >>> signum("a")
     >>> signum([])
```
which should hopefully raise Exceptions.

**Review by cclauss — CHANGES_REQUESTED** [review]
a few more

File: maths/signum.py [hunk 4]
```diff
@@ -26,9 +35,15 @@ def test_signum() -> None:
...
 
 
-if __name__ == "__main__":
+if __ame__ == "__main":
```
Comment (cclauss) [inline maths/signum.py:46]: Please revert this change.

File: maths/signum.py [hunk 5]
```diff
@@ -7,12 +7,21 @@ def signum(num: float) -> int:
...
+    Custom test cases:
+    >>> signum(-20)
     -1
-    >>> signum(10)
+    >>> signum(20)
     1
     >>> signum(0)
     0
+    >>> signum("a")
+    0
```
Comment (cclauss) [inline maths/signum.py:18]:
```suggestion
    Traceback (most recent call last):
        ...
    TypeError: '<' not supported between instances of 'str' and 'int'
```

Comment (imSanko) [reply maths/signum.py]: Check I've added changes

**Comment by cclauss** [comment]
```
 021     >>> signum([])
UNEXPECTED EXCEPTION: TypeError("'<' not supported between instances of 'list' and 'int'")
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.12.0/x64/lib/python3.12/doctest.py", line 1357, in __run
    exec(compile(example.source, filename, "single",
  File "<doctest maths.signum.signum[4]>", line 1, in <module>
  File "/home/runner/work/Python/Python/maths/signum.py", line 28, in signum
    if num < 0:
       ^^^^^^^
TypeError: '<' not supported between instances of 'list' and 'int'
/home/runner/work/Python/Python/maths/signum.py:21: UnexpectedException
```

**Comment by imSanko** [comment]
> ```
>  021     >>> signum([])
> UNEXPECTED EXCEPTION: TypeError("'<' not supported between instances of 'list' and 'int'")
> Traceback (most recent call last):
>   File "/opt/hostedtoolcache/Python/3.12.0/x64/lib/python3.12/doctest.py", line 1357, in __run
>     exec(compile(example.source, filename, "single",
>   File "<doctest maths.signum.signum[4]>", line 1, in <module>
>   File "/home/runner/work/Python/Python/maths/signum.py", line 28, in signum
>     if num < 0:
>        ^^^^^^^
> TypeError: '<' not supported between instances of 'list' and 'int'
> /home/runner/work/Python/Python/maths/signum.py:21: UnexpectedException
> ```

Test cases have passed
python -m doctest -v maths/signum.py 
Trying:
    signum(-20)
Expecting:
    -1
ok
Trying:
    signum(20)
Expecting:
    1
ok
Trying:
    signum(0)
Expecting:
    0
ok
Trying:
    signum("a")
Expecting:
    0
ok
Trying:
    signum([])
Expecting:
    0
ok
Trying:
    signum(-10)
Expecting:
    -1
ok
Trying:
    signum(10)
Expecting:
    1
ok
2 items had no tests:
    signum
    signum.test_signum
1 items passed all tests:
   7 tests in signum.signum
7 tests in 3 items.
7 passed and 0 failed.
Test passed.

File: maths/signum.py [hunk 6]
```diff
@@ -3,20 +3,35 @@
 """
 
 
-def signum(num: float) -> int:
+def signum(num) -> int:
```
Comment (cclauss) [inline maths/signum.py:6]: Why remove a type hint?!?

File: maths/signum.py (hunk 5 above)
Comment (cclauss) [reply maths/signum.py]: Please revert this change.  The code CORRECTLY raises a TypeError if you try to give it a string instead of a number.  This is MUCH better than sweeping that error under the rug and _pretending_ the answer is zero when it is NOT.

File: maths/signum.py [hunk 7]
```diff
@@ -3,20 +3,37 @@
...
-    return 1 if num else 0
+    if isinstance(num, (int, float)):
+        if num < 0:
+            return -1
+        elif num > 0:
+            return 1
+        else:
+            return 0
+    else:
+        return 0
```
Comment (cclauss) [inline maths/signum.py:36]: Revert this change.  We should not change the code to _fit_ the tests.

File: maths/signum.py [hunk 8]
```diff
@@ -3,20 +3,37 @@
...
+    >>> signum(20)
     1
     >>> signum(0)
     0
+    >>> signum("a")
+    Traceback (most recent call last):
+        ...
+    TypeError: '<' not supported between instances of 'str' and 'int'
+    >>> signum([])
+    0
```
Comment (cclauss) [inline maths/signum.py:20]: This should also be a TypeError... https://github.com/TheAlgorithms/Python/pull/10724#issuecomment-1772362052

**Review by cclauss — APPROVED** [review]

**Comment by imSanko** [comment]
@cclauss Thanks a lot for mentoring me

**Comment by cclauss** [comment]
The trick is NOT to change the code to fix the test but to change ONLY the test to describe how the code ACTUALLY works.
